README.md
f.lux indicator applet
======================
Better lighting for your computer
Learn more here: https://justgetflux.com/

f.lux indicator applet is an indicator applet to control xflux: an application
that makes the color of your computer's display adapt to the time of day, warm
at nights and like sunlight during the day.

INSTALLATION FROM SOURCE CODE
--------
###STEP 1 Install the dependencies

Fedora:

```
sudo yum install gnome-python2-gconf pexpect python-appindicator git
```

Debian:

```
sudo apt-get install python-gconf python-pexpect python-appindicator git fakeroot debhelper python-support
```

###STEP 2 Download the source code

```
git clone https://github.com/Kilian/f.lux-indicator-applet.git
cd f.lux-indicator-applet
```

###STEP 3 Download xflux (not required for Debian/Ubuntu)

32bit:

```
wget https://justgetflux.com/linux/xflux-pre.tgz
tar zxvf xflux-pre.tgz
```

64bit:

```
wget https://justgetflux.com/linux/xflux64.tgz
tar zxvf xflux64.tgz
```


###STEP 4 Build and install

Debian/Ubuntu:

```
dpkg-buildpackage -rfakeroot -b -uc -us
sudo dpkg -i ../flux_1.1.8_all.deb (check filename)
```

Other distros:

```
python setup.py build
sudo python setup.py install
cd ..
rm -r f.lux-indicator-applet
```


debian/changelog
fluxgui (1.1.8-4~vivid) vivid; urgency=medium

  * Add debhelper tokens to post{inst,rm}
    - Fixes module not being found by python

 -- Nathan Rennie-Waldock <nathan.renniewaldock@gmail.com>  Mon, 01 Feb 2016 21:00:40 +0000

fluxgui (1.1.8-3~vivid) vivid; urgency=medium

  * Add some text to __init__.py
    - Was being left out package as it was empty

 -- Nathan Rennie-Waldock <nathan.renniewaldock@gmail.com>  Mon, 01 Feb 2016 17:21:20 +0000

fluxgui (1.1.8-2~vivid) vivid; urgency=medium

  * Add missing python-glade2 dependency
  * dh_clean should also remove xflux

 -- Nathan Rennie-Waldock <nathan.renniewaldock@gmail.com>  Mon, 01 Feb 2016 17:03:47 +0000

fluxgui (1.1.8-1~vivid) vivid; urgency=medium

  * Non-maintainer upload

  [ Jeremy Newton ]
  * Remove xflux binary

  [ Nathan Rennie-Waldock ]
  * Download xflux binary automatically on install
  * Update dependencies properly
  * Updated colour temperature choices

 -- Nathan Rennie-Waldock <nathan.renniewaldock@gmail.com>  Sat, 15 Aug 2015 22:56:25 +0100

fluxgui (1.1.8) lucid; urgency=low

  * add in python-pexpect as a dependency 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Wed, 08 Sep 2010 12:12:16 +0200

fluxgui (1.1.7-0ubuntu1) lucid; urgency=low

  * handle xflux closing more gracefully (thanks to axonxorz) 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sat, 31 Jul 2010 15:29:01 +0200

fluxgui (1.1.6-0ubuntu1) lucid; urgency=low

  * make the preview button preview the selected value instead of current
  * update xflux
  * improved KDE support (needs testing)

 -- Kilian Valkhof <kilian@ktop3>  Thu, 29 Jul 2010 14:32:55 +0200

fluxgui (1.1.5-0ubuntu1) lucid; urgency=low

  * remove any stray lock files if xflux can't be started (by seancron) 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 23:15:51 +0200

fluxgui (1.1.4-0ubuntu1) lucid; urgency=low

  * fix problems with the first start of the application when there are no settings yet 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 17:46:41 +0200

fluxgui (1.1.3-0ubuntu1) lucid; urgency=low

  * clean up properly after a keyboardinterrupt 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 14:05:51 +0200

fluxgui (1.1.2-0ubuntu1) lucid; urgency=low

  * add a close button to the preferences pane 

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Mon, 26 Jul 2010 12:08:10 +0200

fluxgui (1.1.1-0ubuntu1) lucid; urgency=low

  * update some debian settings

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Sun, 25 Jul 2010 14:18:19 +0200

fluxgui (1.1.0-0ubuntu1) lucid; urgency=low

  * prevent fluxgui from starting more than once
  * implement a preview button
  * implement an autostart function

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Thu, 15 Jul 2010 21:08:59 +0200

fluxgui (1.0.0-0ubuntu1) lucid; urgency=low

  * gnome indicator applet for xflux
  * own logo
  * starts xflux automatically
  * adapts icon to lucid style
  * allows on/off toggling of xflux
  * saves the latitude/zipcode settings in gconf
  * opens preferences window if latitude/zipcode is not yet set

 -- Kilian Valkhof <kilian@kilianvalkhof.com>  Tue, 15 Jun 2010 22:14:58 +0200



src/fluxgui/fluxapp.py
#!/usr/bin/python2.7

from fluxgui import fluxcontroller, settings
from fluxgui.exceptions import MethodUnavailableError
import gtk
import gtk.glade
import appindicator
import sys, os
import signal
import errno

class FluxGUI(object):
    """
    FluxGUI initializes/destroys the app
    """
    def __init__(self):
        self.pidfile = os.path.expanduser("~/.fluxgui.pid")
        self.check_pid()
        try:
            self.settings = settings.Settings()
            self.xflux_controller = fluxcontroller.FluxController(self.settings)
            self.indicator = Indicator(self, self.xflux_controller)
            self.preferences = Preferences(self.settings,
                    self.xflux_controller)
            self.xflux_controller.start()

        except Exception as e:
            print e
            print "Critical error. Exiting."
            self.exit(1)

    def __del__(self):
        self.exit()

    def open_preferences(self):
        self.preferences.show()

    def signal_exit(self, signum, frame):
        print 'Recieved signal: ', signum
        print 'Quitting...'
        self.exit()

    def exit(self, code=0):
        try:
            self.xflux_controller.stop()
        except MethodUnavailableError:
            pass
        os.unlink(self.pidfile)
        gtk.main_quit()
        sys.exit(code)

    def run(self):
        gtk.main()

    def check_pid(self):
        pid = os.getpid()

        running = False # Innocent...
        if os.path.isfile(self.pidfile):
            try:
                oldpid = int(open(self.pidfile).readline().rstrip())
                try:
                    os.kill(oldpid, 0)
                    running = True # ...until proven guilty
                except OSError as err:
                    if err.errno == errno.ESRCH:
                        # OSError: [Errno 3] No such process
                        print "stale pidfile, old pid: ", oldpid
            except ValueError:
                # Corrupt pidfile, empty or not an int on first line
                pass
        if running:
            print "fluxgui is already running, exiting"
            sys.exit()
        else:
            file(self.pidfile, 'w').write("%d\n" % pid)

class Indicator(object):
    """
    Information and methods related to the indicator applet.
    Executes FluxController and FluxGUI methods.
    """

    def __init__(self, fluxgui, xflux_controller):
        self.fluxgui = fluxgui
        self.xflux_controller = xflux_controller
        self.indicator = appindicator.Indicator(
          "fluxgui-indicator",
          "fluxgui",
          appindicator.CATEGORY_APPLICATION_STATUS)

        self.setup_indicator()

    def setup_indicator(self):
        self.indicator.set_status(appindicator.STATUS_ACTIVE)

        # Check for special Ubuntu themes. copied from lookit

        try:
            theme = \
                gtk.gdk.screen_get_default().get_setting('gtk-icon-theme-name')
        except:
            self.indicator.set_icon('fluxgui')
        else:
            if theme == 'ubuntu-mono-dark':
                self.indicator.set_icon('fluxgui-dark')
            elif theme == 'ubuntu-mono-light':
                self.indicator.set_icon('fluxgui-light')
            else:
                self.indicator.set_icon('fluxgui')

        self.indicator.set_menu(self.create_menu())

    def create_menu(self):
        menu = gtk.Menu()

        self.add_menu_item("_Pause f.lux", self._toggle_pause,
                menu, MenuItem=gtk.CheckMenuItem)
        self.add_menu_item("Prefere_nces", self._open_preferences, menu)
        self.add_menu_separator(menu)
        self.add_menu_item("Quit", self._quit, menu)

        return menu

    def add_menu_item(self, label, handler, menu,
            event="activate", MenuItem=gtk.MenuItem, show=True):
        item = MenuItem(label)
        item.connect(event, handler)
        menu.append(item)
        if show:
            item.show()
        return item

    def add_menu_separator(self, menu, show=True):
        item = gtk.SeparatorMenuItem()
        menu.append(item)
        if show:
            item.show()

    def _toggle_pause(self, item):
        self.xflux_controller.toggle_pause()

    def _open_preferences(self, item):
        self.fluxgui.open_preferences()

    def _quit(self, item):
        self.fluxgui.exit()

class Preferences(object):
    """
    Information and methods related to the preferences window.
    Executes FluxController methods and gets data from Settings.

    """

    temperatureKeys = {
                0:  '2300',
                1:  '2700',
                2:  '3400',
                3:  '4200',
                4:  '5000',
                5:  '6500',
                "off": '6500',
    }

    def temperature_to_key(self, temperature):
        for i, t in self.temperatureKeys.items():
            if t == temperature:
                return i

    def connect_widget(self, widget_name, connect_target=None,
            connect_event="activate"):
        widget = self.wTree.get_widget(widget_name)
        if connect_target:
            widget.connect(connect_event, connect_target)
        return widget


    def __init__(self, settings, xflux_controller):
        self.settings = settings
        self.xflux_controller = xflux_controller

        self.gladefile = os.path.join(os.path.dirname(os.path.dirname(
          os.path.realpath(__file__))), "fluxgui/preferences.glade")
        self.wTree = gtk.glade.XML(self.gladefile)

        self.window = self.connect_widget("window1", self.delete_event,
                connect_event="destroy")
        self.latsetting = self.connect_widget("entryLatitude",
                self.delete_event)
        self.lonsetting = self.connect_widget("entryLongitude",
                self.delete_event)
        self.zipsetting = self.connect_widget("entryZipcode",
                self.delete_event)
        self.colsetting = self.connect_widget("comboColor")
        self.colordisplay = self.connect_widget("labelCurrentColorTemperature")
        self.previewbutton = self.connect_widget("buttonPreview",
                self.preview_click_event, "clicked")
        self.closebutton = self.connect_widget("buttonClose",
                self.delete_event, "clicked")
        self.autostart = self.connect_widget("checkAutostart")

        if (self.settings.latitude is "" and self.settings.zipcode is "")\
                or not self.settings.has_set_prefs:
            self.show()
            self.display_no_zipcode_or_latitude_error_box()

    def show(self):

        self.latsetting.set_text(self.settings.latitude)
        self.lonsetting.set_text(self.settings.longitude)
        self.zipsetting.set_text(self.settings.zipcode)
        self.colsetting.set_active(self.temperature_to_key(self.settings.color))
        self.colordisplay.set_text("Current color temperature: %sK"
                                    % (self.settings.color))
        if self.settings.autostart:
            self.autostart.set_active(True)
        else:
            self.autostart.set_active(False)

        self.window.show()

    def display_no_zipcode_or_latitude_error_box(self):
        md = gtk.MessageDialog(self.window,
                gtk.DIALOG_DESTROY_WITH_PARENT, gtk.MESSAGE_INFO,
                gtk.BUTTONS_OK, "The f.lux indicator applet needs to know " +
                "your latitude or zipcode to run. " +
                "Please fill either of them in on the preferences screen "+
                "and click 'Close'.")
        md.set_title("f.lux indicator applet")
        md.run()
        md.destroy()

    def preview_click_event(self, widget, data=None):
        colsetting_temperature = self.temperatureKeys[
            self.colsetting.get_active()]
        self.xflux_controller.preview_color(colsetting_temperature)

    def delete_event(self, widget, data=None):
        if self.settings.latitude != self.latsetting.get_text():
            self.xflux_controller.set_xflux_latitude(
                    self.latsetting.get_text())

        if self.settings.longitude != self.lonsetting.get_text():
            self.xflux_controller.set_xflux_longitude(
                    self.lonsetting.get_text())

        if self.settings.zipcode != self.zipsetting.get_text():
            self.xflux_controller.set_xflux_zipcode(
                    self.zipsetting.get_text())

        colsetting_temperature = self.temperatureKeys[
                self.colsetting.get_active()]
        if self.settings.color != colsetting_temperature:
            self.xflux_controller.color = colsetting_temperature

        if self.autostart.get_active():
            self.xflux_controller.set_autostart(True)
        else:
            self.xflux_controller.set_autostart(False)
        if self.latsetting.get_text() == "" \
                and self.zipsetting.get_text() == "":
            self.display_no_zipcode_or_latitude_error_box()
            return True

        self.window.hide()
        return False


if __name__ == '__main__':
    try:
        app = FluxGUI()
        signal.signal(signal.SIGTERM, app.signal_exit)
        app.run()
    except KeyboardInterrupt:
        app.exit()



src/fluxgui/settings.py
import os
import gconf
from xdg.DesktopEntry import DesktopEntry
from fluxgui.exceptions import DirectoryCreationError


class Settings(object):

    def __init__(self):
        self.client = GConfClient("/apps/fluxgui")

        self._color = self.client.get_client_string("colortemp", 3400)
        self._autostart = self.client.get_client_bool("autostart")
        self._latitude = self.client.get_client_string("latitude")
        self._longitude = self.client.get_client_string("longitude")
        self._zipcode = self.client.get_client_string("zipcode")

        self.has_set_prefs = True
        if not self._latitude and not self._zipcode:
            self.has_set_prefs = False
            self._zipcode = '90210'
            self.autostart=True
        if int(self._color) < 1200 or not self._color:
            # upgrade from previous version
            temperature_keys = {
                    '0':  '2700',
                    '1':  '3400',
                    '2':  '4200',
                    '3':  '5000',
                    '4':  '6500',
            }
            if self._color in temperature_keys:
                self.color = temperature_keys[self._color]
            else:
                self.color = '3400'

    def xflux_settings_dict(self):
        d = {
                'color': self.color,
                'latitude': self.latitude,
                'longitude': self.longitude,
                'zipcode': self.zipcode,
                'pause_color': '6500'
        }
        return d

    def _get_color(self):
        return str(self._color)
    def _set_color(self, value):
        self._color = value
        self.client.set_client_string("colortemp", value)

    def _get_latitude(self):
        return str(self._latitude)
    def _set_latitude(self, value):
        self._latitude = value
        self.client.set_client_string("latitude", value)

    def _get_longitude(self):
        return str(self._longitude)
    def _set_longitude(self, value):
        self._longitude = value
        self.client.set_client_string("longitude", value)

    def _get_zipcode(self):
        return str(self._zipcode)
    def _set_zipcode(self, value):
        self._zipcode = value
        self.client.set_client_string("zipcode", value)

    def _get_autostart(self):
        return bool(self._autostart)
    def _set_autostart(self, value):
        self._autostart = value
        self.client.set_client_bool("autostart", self._autostart)
        if self._autostart:
            self._create_autostarter()
        else:
            self._delete_autostarter()

    color=property(_get_color, _set_color)
    latitude=property(_get_latitude, _set_latitude)
    longitude=property(_get_longitude, _set_longitude)
    zipcode=property(_get_zipcode, _set_zipcode)
    autostart=property(_get_autostart, _set_autostart)


    #autostart code copied from AWN
    def _get_autostart_file_path(self):
        autostart_dir = os.path.join(os.environ['HOME'], '.config',
                                     'autostart')
        return os.path.join(autostart_dir, 'fluxgui.desktop')

    def _create_autostarter(self):
        autostart_file = self._get_autostart_file_path()
        autostart_dir = os.path.dirname(autostart_file)

        if not os.path.isdir(autostart_dir):
            #create autostart dir
            try:
                os.mkdir(autostart_dir)
            except DirectoryCreationError, e:
                print "Creation of autostart dir failed, please make it yourself: %s" % autostart_dir
                raise e

        if not os.path.isfile(autostart_file):
            #create autostart entry
            starter_item = DesktopEntry(autostart_file)
            starter_item.set('Name', 'f.lux indicator applet')
            starter_item.set('Exec', 'fluxgui')
            starter_item.set('Icon', 'fluxgui')
            starter_item.set('X-GNOME-Autostart-enabled', 'true')
            starter_item.write()
            self.autostart = True

    def _delete_autostarter(self):
        autostart_file = self._get_autostart_file_path()
        if os.path.isfile(autostart_file):
            os.remove(autostart_file)
            self.autostart = False

class GConfClient(object):
    """
    Gets and sets gconf settings.
    """

    def __init__(self, prefs_key):
        self.client = gconf.client_get_default()
        self.prefs_key = prefs_key
        self.client.add_dir(self.prefs_key, gconf.CLIENT_PRELOAD_NONE)

    def get_client_string(self, property_name, default=""):
        client_string = self.client.get_string(self.prefs_key+"/"+property_name)
        if client_string is None:
            client_string = default
        return client_string

    def set_client_string(self, property_name, value):
        self.client.set_string(self.prefs_key + "/" + property_name, str(value))

    def get_client_bool(self, property_name, default=True):
        try:
            gconf_type = self.client.get(self.prefs_key + "/"
                                            + property_name).type
        except AttributeError:
            # key is not set
            self.set_client_bool(property_name, default)
            client_bool = default
            return client_bool

        client_bool = None
        if gconf_type != gconf.VALUE_BOOL:
            # previous release used strings for autostart, handle here
            client_string = self.get_client_string(property_name).lower()
            if client_string == '1':
                self.set_client_bool(property_name, True)
                client_bool = True
            elif client_string == '0':
                self.set_client_bool(property_name, False)
                client_bool = False
        else:
            client_bool = self.client.get_bool(self.prefs_key
                                        + "/"+property_name)
        return client_bool

    def set_client_bool(self, property_name, value):
        self.client.set_bool(self.prefs_key + "/" + property_name, bool(value))

